### PR TITLE
[stdlib] Take several underscored stdlib functions private

### PR DESCRIPTION
**Pull request by airspeedswift**
In many cases, these have leaked out unnecessarily into the SDK or tests. In a few cases, they're needed to do the tests, and might need some work to test in different ways. Mostly just testing perf and compat impact for now.

Starting with various assertiony things.

### CODE REVIEW

**Comment by airspeedswift** [comment]
@swift-ci please smoke benchmark

**Comment by airspeedswift** [comment]
@swift-ci please test source compatibility

**Comment by airspeedswift** [comment]
@swift-ci please smoke test compiler performance

File: stdlib/public/SDK/Foundation/NSArray.swift [hunk 1]
```diff
@@ -31,7 +31,7 @@ extension Array : _ObjectiveCBridgeable {
   /// The provided `NSArray` will be copied to ensure that the copy can
   /// not be mutated by other code.
   internal init(_cocoaArray: NSArray) {
-    _sanityCheck(_isBridgedVerbatimToObjectiveC(Element.self),
+    precondition(_isBridgedVerbatimToObjectiveC(Element.self),
```
Comment (jrose-apple) [inline stdlib/public/SDK/Foundation/NSArray.swift]: This seems like an expensive upgrade (for most of the Foundation things). `assert` is probably fine, given that `_sanityCheck` would have been compiled out in a release build too.

File: test/SILOptimizer/specialize_checked_cast_branch.swift [hunk 2]
```diff
@@ -34,8 +34,6 @@ public func ArchetypeToArchetypeCast<T1, T2>(t1 : T1, t2 : T2) -> T2 {
 // CHECK:   return [[T0]]
 //
 // CHECK: bb2
-// CHECK:   integer_literal $Builtin.Int1, -1
-// CHECK:   cond_fail
```
Comment (eeckstein) [inline test/SILOptimizer/specialize_checked_cast_branch.swift]: Please please the integer_literal+cond_fail with a check for the trap builtin:

// CHECK:   builtin "int_trap"

**Review by eeckstein — CHANGES_REQUESTED** [review]

File: stdlib/public/SDK/Foundation/NSArray.swift (hunk 1 above)
Comment (airspeedswift) [reply stdlib/public/SDK/Foundation/NSArray.swift]: yeah, that's the better call.

**Comment by airspeedswift** [comment]
@swift-ci please test

File: test/SILOptimizer/specialize_checked_cast_branch.swift (hunk 2 above)
Comment (airspeedswift) [reply test/SILOptimizer/specialize_checked_cast_branch.swift]: Done

**Comment by airspeedswift** [comment]
Test with apple/swift-corelibs-foundation#1640

@swift-ci please test linux platform

**Comment by airspeedswift** [comment]
@swift-ci please test macOS platform

**Comment by airspeedswift** [comment]
Test with apple/swift-corelibs-foundation#1640

@swift-ci please test linux platform

**Comment by airspeedswift** [comment]
@swift-ci please test macOS platform

**Comment by airspeedswift** [comment]
@swift-ci please test macOS platform

**Comment by airspeedswift** [comment]
Test with apple/swift-corelibs-foundation#1640

@swift-ci please test linux platform

**Comment by airspeedswift** [comment]
@swift-ci please test macOS platform

**Comment by airspeedswift** [comment]
@swift-ci clean test macOS platform

**Pull request merged**